File: FingoDriver/src/main/java/com/yelloco/fingodriver/repository/FingoPresenter.java
package com.yelloco.fingodriver.repository;

import android.app.Activity;

import com.yelloco.fingodriver.callbacks.FingoContract;
import com.yelloco.fingodriver.enums.Currency;
import com.yelloco.fingodriver.enums.FingoOperation;
import com.yelloco.fingodriver.models.FingoModel;
import com.yelloco.fingodriver.models.fingo_operation.DisplayTextRequested;
import com.yelloco.fingodriver.models.fingo_operation.IdentifyData;
import com.yelloco.fingodriver.models.fingo_operation.PaymentData;
import com.yelloco.fingodriver.models.fingo_operation.ProcessingFinished;
import com.yelloco.fingodriver.models.networking.fingo_error.FingoErrorResponse;
import com.yelloco.fingodriver.models.networking.payment.PosData;
import com.yelloco.fingodriver.models.networking.refund.TerminalData;

public class FingoPresenter implements FingoContract.Presenter
{
    // Members
    private FingoContract.FingoListener fingoListener;
    private FingoContract.Model model;
    private Activity activity;

    public FingoPresenter(Activity activity, FingoContract.FingoListener fingoListener){
        this.activity = activity;
        this.fingoListener = fingoListener;
        this.model = new FingoModel(this, activity);
    }

    public void identify(int timeoutInMillis){
        execute(FingoOperation.IDENTIFY, timeoutInMillis);
    }

    public void enroll(int timeoutInMillis) {
        execute(FingoOperation.ENROLLMENT, timeoutInMillis);
    }

    public void payment(int totalAmount, Currency currency, int totalDiscount, PosData posData, int timeoutInMillis){
        new Thread(() -> {
            this.model.invoke(FingoOperation.PAYMENT);
            if(! this.model.isOperationCancelled()){
                this.model.payment(totalAmount, currency, totalDiscount, posData, timeoutInMillis);
            }
        }).start();
    }

    public void refund(int refundAmount, String transactionIdToRefund, String gatewayTransactionIdToRefund, TerminalData terminalData, int timeoutInMillis){
        new Thread(() -> {
            this.model.invoke(FingoOperation.REFUND);
            if(! this.model.isOperationCancelled()){
                this.model.refund(refundAmount, transactionIdToRefund, gatewayTransactionIdToRefund, terminalData, timeoutInMillis);
            }
        }).start();
    }

    @Override
    public void cancel() {
        new Thread(() -> {
            this.model.cancel();
        }).start();
    }

    @Override
    public boolean isDeviceConnected(){
        return this.model.isDeviceConnected();
    }

    @Override
    public boolean isOperationCancelled() {
        return this.model.isOperationCancelled();
    }

    private void execute(FingoOperation fingoOperation, int timeoutInMillis){
        new Thread(() -> {
            this.model.invoke(fingoOperation);
            if(! this.model.isOperationCancelled()){
                switch (fingoOperation){
                    case IDENTIFY:{
                        this.model.identify(timeoutInMillis);
                        break;
                    }
                    case ENROLLMENT:{
                        this.model.enroll(timeoutInMillis);
                        break;
                    }
                }
            }
        }).start();
    }

    @Override
    public void onProcessingStarted() {
        this.fingoListener.onProcessingStarted();
    }

    @Override
    public void onDisplayTextRequested(DisplayTextRequested displayTextRequested){
        this.fingoListener.onDisplayTextRequested(displayTextRequested);
    }

    @Override
    public void onIdentifyData(IdentifyData identifyData) {
        this.fingoListener.onIdentifyData(identifyData);
    }

    @Override
    public void onPaymentData(PaymentData paymentData, FingoErrorResponse fingoErrorResponse) {
        this.fingoListener.onPaymentData(paymentData, fingoErrorResponse);
    }

    @Override
    public void onProcessingFinished(ProcessingFinished processingFinished) {
        activity.runOnUiThread(() -> {
            this.fingoListener.onProcessingFinished(processingFinished);
        });
    }
}

File: FingoDriver/src/main/java/com/yelloco/fingodriver/repository/FingoPresenter.kt
package com.yelloco.fingodriver.repository

import android.content.Context
import android.util.Log
import com.yelloco.fingodriver.FingoFactory
import com.yelloco.fingodriver.callbacks.FingoContract
import com.yelloco.fingodriver.enums.Currency
import com.yelloco.fingodriver.enums.FingoOperation
import com.yelloco.fingodriver.models.FingoModel
import com.yelloco.fingodriver.models.fingo_operation.IdentifyData
import com.yelloco.fingodriver.models.fingo_operation.PaymentData
import com.yelloco.fingodriver.models.fingo_operation.ProcessingFinished
import com.yelloco.fingodriver.models.fingo_operation.display_text_requested.DisplayTextRequested
import com.yelloco.fingodriver.models.networking.fingo_error.FingoErrorResponse
import com.yelloco.fingodriver.models.networking.payment.PosData
import com.yelloco.fingodriver.models.networking.refund.TerminalData
import kotlinx.coroutines.launch

class FingoPresenter(
    context: Context, // Members
    private val fingoListener: FingoContract.FingoListener
) : FingoContract.Presenter
{
    private val model: FingoContract.Model

    init {
        model = FingoModel(this, context)
    }

    fun identify(timeoutInMillis: Long) {
        execute(FingoOperation.IDENTIFY, timeoutInMillis)
    }

    fun enroll(timeoutInMillis: Long) {
        execute(FingoOperation.ENROLLMENT, timeoutInMillis)
    }

    fun payment(
        totalAmount: Int,
        currency: Currency?,
        totalDiscount: Int,
        posData: PosData?,
        timeoutInMillis: Long
    ) {
        FingoFactory.IO_SCOPE.launch {
            launch {
                model.invoke(FingoOperation.PAYMENT)
                if (!model.isOperationCancelled) {
                    model.payment(
                        totalAmount,
                        currency,
                        totalDiscount,
                        posData,
                        timeoutInMillis
                    )
                }
            }.invokeOnCompletion {
                if(it == null){
                    Log.d(TAG, "Payment Job completed successfully")
                }
                else{
                    Log.d(TAG, "Payment Job Failed")
                }
            }
        }
    }

    fun refund(
        refundAmount: Int,
        transactionIdToRefund: String?,
        gatewayTransactionIdToRefund: String?,
        terminalData: TerminalData?,
        timeoutInMillis: Long
    ) {
        FingoFactory.IO_SCOPE.launch {
            launch {
                model.invoke(FingoOperation.REFUND)
                if (!model.isOperationCancelled) {
                    model.refund(
                        refundAmount,
                        transactionIdToRefund,
                        gatewayTransactionIdToRefund,
                        terminalData,
                        timeoutInMillis
                    )
                }
            }.invokeOnCompletion {
                if(it == null){
                    Log.d(TAG, "Refund Job completed successfully")
                }
                else{
                    Log.d(TAG, "Refund Job Failed")
                }
            }
        }
    }

    override fun cancel() {
        Thread { model.cancel() }.start()
    }

    override val isDeviceConnected: Boolean
        get() = model.isDeviceConnected
    override val isOperationCancelled: Boolean
        get() = model.isOperationCancelled

    private fun execute(fingoOperation: FingoOperation, timeoutInMillis: Long) {
        FingoFactory.IO_SCOPE.launch {
            launch {
                model.invoke(fingoOperation)
                if (!model.isOperationCancelled) {
                    when (fingoOperation) {
                        FingoOperation.IDENTIFY -> {
                            model.identify(timeoutInMillis)
                        }
                        FingoOperation.ENROLLMENT -> {
                            model.enroll(timeoutInMillis)
                        }
                    }
                }
            }.invokeOnCompletion {
                if(it == null){
                    Log.d(TAG, "${fingoOperation.name} Job completed successfully")
                }
                else{
                    Log.d(TAG, "${fingoOperation.name} Job Failed")
                }
            }
        }
    }

    override fun onProcessingStarted() {
        fingoListener.onProcessingStarted()
    }

    override fun onDisplayTextRequested(displayTextRequested: DisplayTextRequested?) {
        fingoListener.onDisplayTextRequested(displayTextRequested)
    }

    override fun onIdentifyData(identifyData: IdentifyData?) {
        fingoListener.onIdentifyData(identifyData)
    }

    override fun onPaymentData(paymentData: PaymentData?, fingoErrorResponse: FingoErrorResponse?) {
        fingoListener.onPaymentData(paymentData, fingoErrorResponse)
    }

    override fun onProcessingFinished(processingFinished: ProcessingFinished?) {
        fingoListener.onProcessingFinished(processingFinished)
    }

    companion object {
        private const val TAG = "FingoPresenter"
    }
}